sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memcpy.asciidoc

```diff
@@ -0,0 +1,161 @@
+permission by Khronos.
+
+
+== Contact
+
+To report problems with this extension, please open a new issue at:
+
+https://github.com/intel/llvm/issues
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 revision 7 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+This extension also depends on the following SYCL extensions:
+
+* link:https://github.com/intel/llvm/pull/9186/[sycl_ext_oneapi_barrier]
+
+== Status
+
+This is a proposed extension specification, intended to gather community
+feedback.  Interfaces defined in this specification may not be implemented yet
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+This extension defines
+`sycl::ext::oneapi::experimental::async_memcpy` free function to
+generalize and replace the current `sycl::async_work_group_copy`
+function. 
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ASYNC_MEMCPY` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== `async_memcpy` function
+`sycl::ext::oneapi::experimental::async_memcpy` is a free function
+that  asynchronously copies a number of elements specified by
+`num_elements` of data of type `T` from the source pointer `src` to
+destination pointer `dest`. It also takes a barrier object of type
+`syclex::barrier` as an argument that can be used to wait on the
+completion of the memory copy.
+
+Permitted types for `T` are all scalar and vector types.
+
+This extension provides two versions of `async_memcpy`: with and
+without `Group` template parameter and argument. In the case of the
+group variant, `group_async_memcpy` is issued by all the threads in
+the group. This is a _group function_, as defined in Section 4.17.3
+of the SYCL specification. In the case of the work-item variant,
+`async_memcpy` is issued by the current work-item.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  template <typename T,  access::address_space DestSpace,
+  access::decorated DestIsDecorated, access::address_space SrcSpace,
+  access::decorated SrcIsDecorated, sycl::memory_scope Scope>
+  void async_memcpy(multi_ptr<T, DestSpace, DestIsDecorated> dest,
```

Comment:
Why limit to `multi_ptr` here?  Should we allow generic `T*` as well?

Reply:
I use `multi_ptr` to be able to carry around the address space info.
If USM is used, `multi_ptr` can be constructed using `address_space_cast`
Can the implementation of `async_copy` retrieve address space (local vs global) if we only pass generic `T*`?

Reply:
It depends.  `T*` is allowed to match both decorated pointers (which have the address space attached) and raw pointers (which don't have the address space attached).

If a function is passed a decorated pointer, then the implementation knows the address space (just as if it was passed a `multi_ptr`).  If it's passed a raw pointer, then it can use `address_space_cast` to determine what the address space is using a runtime check (which may introduce some overhead).

I'm hopeful that in cases where the compiler actually does know the address space, but it's represented as a raw pointer, the runtime check can be optimized away.  But I don't think that is implemented right now.